frontend/src/components/grantaccess/GrantAccess.vue

```diff
@@ -234,94 +235,103 @@ function toRequestPayload(
     return request;
 }
 
+const composeAndPushNotificationMessage = (
+    severity: Severity,
+    clientIDList: string[]
+) => {
+    let username = formData.value.userId.toUpperCase();
+    if (severity === Severity.success) {
+        pushNotification(
+            severity,
+            `${username} ${
+                clientIDList[0] === ''
+                    ? `was successfully added with the role ${
+                          getSelectedRole()?.role_name
+                      }`
+                    : `was successfully added with Client IDs: ${clientIDList.join(
+                          ', '
+                      )}`
+            }`
+        );
+    }
+    if (severity === Severity.warning) {
+        pushNotification(
+            severity,
+            `${username} ${
+                clientIDList[0] === ''
+                    ? `already exists with the role ${
+                          getSelectedRole()?.role_name
+                      }`
+                    : `already exists with Client IDs: ${clientIDList.join(
+                          ', '
+                      )}`
+            }`
+        );
+    }
+    if (severity === Severity.error) {
+        pushNotification(
+            severity,
+            `An error has occured. ${username} ${
+                clientIDList[0] === ''
+                    ? `could not be added with the role ${
+                          getSelectedRole()?.role_name
+                      }`
+                    : `was not added with Client IDs: ${clientIDList.join(
+                          ', '
+                      )}`
+            }`
+        );
+    }
+    return '';
+};
+
 async function handleSubmit() {
-    if (forestClientData.value.length > 0) {
-        const successList: string[] = [];
-        const warningList: string[] = [];
-        const errorList: string[] = [];
-        for (const item of forestClientData.value) {
-            const data = toRequestPayload(formData.value, item);
-
-            await userRoleAssignmentApi
-                .createUserRoleAssignment(data as FamUserRoleAssignmentCreate)
-                .then(() => {
-                    successList.push(item.forest_client_number);
-                })
-                .catch((error) => {
-                    if (error.response?.status === 409) {
-                        warningList.push(item.forest_client_number);
-                    } else {
-                        errorList.push(item.forest_client_number);
-                    }
-                });
-        }
+    const successList: string[] = [];
+    const warningList: string[] = [];
+    const errorList: string[] = [];
+
+    do {
+        let item = forestClientData.value.pop() as FamForestClient | null;
```

Comment:
I guess I was not clear on my question.
Just wondering about the "null" case.

Reply:
The `null` is for the cases where the forest client id list is empty, so the same code works for the two scenarios (abstract role or not). Previously there was some sort of code duplication to handle each scenario separately. Also that's why it is doing a `do-while` loop for the list of forest client ids, because it needs to run at least once to call the backend API to assign the role to the user (for the cases where the forest client id is `null`).

Reply:
I see.
Maybe can do this, no need for `as` some type since your `ref` already infers a type:
```
        const item = forestClientData.value.pop();
        const data = toRequestPayload(formData.value, item);
```
and use `undefined` for the `toRequestPayload` function:
```
function toRequestPayload(
    formData: any,
    forestClientData: FamForestClient | undefined
) 
```
see if it works.

Reply:
It does work. Thank you for the suggestion. I'll update the code.